### PR TITLE
MINOR: Pass a streams config to replace the single state dir

### PR DESCRIPTION
**Pull request by guozhangwang**
This is a general change and is re-requisite to allow streams benchmark test with different streams tests. For the streams benchmark itself I will have a separate PR for switching configs. Details:

1. Create a "streams.properties" file under PERSISTENT_ROOT before all the streams test. For now it will only contain a single config of state.dir pointing to PERSISTENT_ROOT.
2. For all the system test related code, replace the main function parameter of state.dir with propsFilename, then inside the function load the props from the file and apply overrides if necessary.
3. Minor fixes.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
@mjsax @bbejeck @vvcephei @dguy for reviews.

**Comment by bbejeck** [comment]
@guozhangwang failures are related, unused import

**Comment by guozhangwang** [comment]
Yup, I'm fixing.

**Comment by guozhangwang** [comment]
Triggered system tests: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1471 Succeeded.

File: streams/src/test/java/org/apache/kafka/streams/tests/BrokerCompatibilityTest.java [hunk 1]
```diff
@@ -47,20 +47,16 @@
...
         final String kafka = args.length > 0 ? args[0] : "localhost:9092";
-        final String stateDirStr = args.length > 1 ? args[1] : TestUtils.tempDirectory().getAbsolutePath();
+        final String propFileName = args.length > 1 ? args[1] : null;
         final boolean eosEnabled = args.length > 2 ? Boolean.parseBoolean(args[2]) : false;
 
-        final File stateDir = new File(stateDirStr);
-        stateDir.mkdir();
-
-        final Properties streamsProperties = new Properties();
+        final Properties streamsProperties = Utils.loadProps(propFileName);
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/tests/BrokerCompatibilityTest.java:57]: `propFileName` could be null, do a null check here?

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsBrokerDownResilienceTest.java [hunk 2]
```diff
@@ -43,17 +44,17 @@
...
 
         final String kafka = args.length > 0 ? args[0] : "localhost:9092";
-        final String stateDirStr = args.length > 1 ? args[1] : TestUtils.tempDirectory().getAbsolutePath();
+        final String propFileName = args.length > 1 ? args[1] : null;
         final String additionalConfigs = args.length > 2 ? args[2] : null;
 
         final Serde<String> stringSerde = Serdes.String();
 
-        final Properties streamsProperties = new Properties();
+        final Properties streamsProperties = Utils.loadProps(propFileName);
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsBrokerDownResilienceTest.java:57]: same here, check for null `propFileName`

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsSmokeTest.java [hunk 3]
```diff
@@ -16,27 +16,32 @@
...
-    public static void main(String[] args) throws InterruptedException {
-        String kafka = args[0];
-        String stateDir = args.length > 1 ? args[1] : null;
-        String command = args.length > 2 ? args[2] : null;
+    public static void main(String[] args) throws InterruptedException, IOException {
+        final String kafka = args[0];
+        final String propFileName = args.length > 1 ? args[1] : null;
+        final String command = args.length > 2 ? args[2] : null;
+
+        final Properties streamsProperties = Utils.loadProps(propFileName);
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsSmokeTest.java:39]: same here

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java [hunk 4]
```diff
@@ -32,34 +33,33 @@
...
         final String kafka = args.length > 0 ? args[0] : "localhost:9092";
-        final String stateDirStr = args.length > 1 ? args[1] : TestUtils.tempDirectory().getAbsolutePath();
+        final String propFileName = args.length > 1 ? args[1] : null;
         final String additionalConfigs = args.length > 2 ? args[2] : null;
 
         final Serde<String> stringSerde = Serdes.String();
 
-        final Properties streamsProperties = new Properties();
-        streamsProperties.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, kafka);
+        final Properties streamsProperties = Utils.loadProps(propFileName);
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java:55]: as above

**Comment by bbejeck** [comment]
Thanks @guozhangwang left a few minor comments, otherwise LGTM.

One general question and sorry if I missed something, but where do we supply our own properties file or edit the existing one to add additional settings?

**Comment by guozhangwang** [comment]
> One general question and sorry if I missed something, but where do we supply our own properties file or edit the existing one to add additional settings?

Thanks for the comments. For your question, the file is added here: 

https://github.com/apache/kafka/pull/4714/files#diff-1dcacbd6daa49466dc7cb89022929dfbR143

that will only encode the state dir for now. In the follow-up PR I will modify simple benchmark python code to allow for other config overridden and remove the hard-coded ones in the java code. Does that make sense?

File: streams/src/test/java/org/apache/kafka/streams/tests/BrokerCompatibilityTest.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/tests/BrokerCompatibilityTest.java:57]: I will consider adding the check within the function.

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java [hunk 5]
```diff
@@ -505,11 +504,15 @@ public static String formatBytes(long bytes) {
...
      * @param filename The path of the file to read
      */
-    public static Properties loadProps(String filename) throws IOException, FileNotFoundException {
+    public static Properties loadProps(String filename) throws IOException {
         Properties props = new Properties();
-        try (InputStream propStream = new FileInputStream(filename)) {
-            props.load(propStream);
+
+        if (filename != null) {
+            try (InputStream propStream = new FileInputStream(filename)) {
```
Comment (mjsax) [inline clients/src/main/java/org/apache/kafka/common/utils/Utils.java:511]: Should we log something for this case?

File: streams/src/test/java/org/apache/kafka/streams/perf/SimpleBenchmark.java [hunk 6]
```diff
@@ -193,34 +195,42 @@ private void run() {
...
+        final String kafka = args.length > 0 ? args[0] : "localhost:9092";
+        final String propFileName = args.length > 1 ? args[1] : null;
+        final int numRecords = args.length > 2 ? Integer.parseInt(args[2]) : 10000000;
+        final boolean loadPhase = args.length > 3 ? Boolean.parseBoolean(args[3]) : false;
+        final String testName = args.length > 4 ? args[4].toLowerCase(Locale.ROOT) : ALL_TESTS;
+        final int numThreads = args.length > 5 ? Integer.parseInt(args[5]) : 1;
+
+        final Properties props = Utils.loadProps(propFileName);
+
+        String stateDirStr;
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/perf/SimpleBenchmark.java]: nit: add `final`

File: streams/src/test/java/org/apache/kafka/streams/perf/SimpleBenchmark.java [hunk 7]
```diff
@@ -193,34 +195,42 @@ private void run() {
...
         System.out.println("loadPhase=" + loadPhase);
         System.out.println("testName=" + testName);
         System.out.println("numThreads=" + numThreads);
 
-        SimpleBenchmark benchmark = new SimpleBenchmark(stateDir, kafka, loadPhase, testName, numRecords, numThreads);
+        SimpleBenchmark benchmark = new SimpleBenchmark(props, kafka, loadPhase, testName, numRecords, numThreads);
         benchmark.run();
     }
 
     public Properties setStreamProperties(final String applicationId) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/perf/SimpleBenchmark.java]: should we change the return type to void?

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java [hunk 8]
```diff
@@ -35,26 +35,25 @@
...
 
     private final String kafka;
-    private final File stateDir;
+    private final Properties streamsProperties;
     private KafkaStreams streams;
     private Thread thread;
     private boolean uncaughtException = false;
 
-    public SmokeTestClient(File stateDir, String kafka) {
+    public SmokeTestClient(Properties streamsProperties, String kafka) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java]: nit: add `final` to cleanup code

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java [hunk 9]
```diff
@@ -94,10 +93,8 @@ public void close() {
         }
     }
 
-    private static KafkaStreams createKafkaStreams(File stateDir, String kafka) {
-        final Properties props = new Properties();
+    private static KafkaStreams createKafkaStreams(Properties props, String kafka) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java]: nit: add `final`

File: streams/src/test/java/org/apache/kafka/streams/perf/SimpleBenchmark.java [hunk 10]
```diff
@@ -193,34 +195,42 @@ private void run() {
...
         System.out.println("kafka=" + kafka);
-        System.out.println("stateDir=" + stateDir);
+        System.out.println("streamsProperties=" + props);
         System.out.println("numRecords=" + numRecords);
         System.out.println("loadPhase=" + loadPhase);
         System.out.println("testName=" + testName);
         System.out.println("numThreads=" + numThreads);
 
-        SimpleBenchmark benchmark = new SimpleBenchmark(stateDir, kafka, loadPhase, testName, numRecords, numThreads);
+        SimpleBenchmark benchmark = new SimpleBenchmark(props, kafka, loadPhase, testName, numRecords, numThreads);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/perf/SimpleBenchmark.java:228]: should we put `kafka` and `numThread` directly into `props` and reduce number of parameters here (would also simplify all the "passing through" code.

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsSmokeTest.java [hunk 11]
```diff
@@ -16,27 +16,32 @@
...
+     *  args ::= kafka propFileName command
      *  command := "run" | "process"
      *
      * @param args
      */
-    public static void main(String[] args) throws InterruptedException {
-        String kafka = args[0];
-        String stateDir = args.length > 1 ? args[1] : null;
-        String command = args.length > 2 ? args[2] : null;
+    public static void main(String[] args) throws InterruptedException, IOException {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsSmokeTest.java]: nit: add `final`

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java [hunk 12]
```diff
@@ -32,36 +33,41 @@
...
+import java.io.IOException;
 import java.util.Map;
 import java.util.Properties;
 import java.util.Set;
 import java.util.concurrent.TimeUnit;
 
 public class StreamsStandByReplicaTest {
 
-    public static void main(String[] args) {
+    public static void main(String[] args) throws IOException {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java]: nit: add `final`

**Review by mjsax — COMMENTED** [review]
Thanks for the PR. Looks good overall. Couple of minor comments.

Can you also run streams system test before we merge this. Thx.

**Comment by guozhangwang** [comment]
> Can you also run streams system test before we merge this. Thx.

Yeah I have triggered a couple streams system test, the first successful one is https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1471

File: streams/src/test/java/org/apache/kafka/streams/perf/SimpleBenchmark.java (hunk 10 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/perf/SimpleBenchmark.java:228]: Yes, in my follow-up PR I've done that, just for its scope I tend to keep its changes more focused.

**Comment by guozhangwang** [comment]
Retriggered https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1480/

**Review by mjsax — APPROVED** [review]
LGTM.

Is there already a follow up PR that address the pending comments?

**Comment by guozhangwang** [comment]
I'll create the follow-up PR once I rebased this one, there are a sequence of them on my local repo :)

**Pull request merged**